birdvoxpaint/util.py

```diff
@@ -0,0 +1,225 @@
+import datetime
+import numpy as np
+
+import librosa
+from librosa.util.exceptions import ParameterError
+
+import joblib
+
+
+sample2frame = lambda x, window, hop: int((x - window) / hop) + 1
+frame2sample = lambda x, window, hop: int((x - 1) * hop) + window
+
+def block_stream(filename=None, y=None, sr=None,
+                 frame_length=2048, hop_length=512,
+                 segment_duration=1, n_blocks=1, full_frames=True):
+    '''Load audio into frames. If given a filename, it will load from file in blocks.
+    If y and sr are given, slice into blocks
+
+    Arguments:
+        filename (str, optional):
+        y, sr (np.ndarray, int, optional):
+        frame_length (int):
+        hop_length (int):
+        block_duration (float)
+
+    '''
+    # load audio frame generator
+
+    if filename is not None:
+        if y is not None:
+            raise ParameterError(
+                'Either y or filename must be equal to None')
+
+        # get blocks from file
+
+        orig_sr = librosa.get_samplerate(filename)
+        sr = sr or orig_sr
+
+        # see: https://librosa.github.io/librosa/_modules/librosa/core/audio.html#stream
+        # block_length is in units of `frames` so reverse calculation
+        block_length = max(segment_duration * orig_sr, frame_length) * n_blocks
+        block_n_frames = sample2frame(block_length, frame_length, hop_length)
+        y_blocks = librosa.stream(filename,
+            block_length=block_n_frames,
+            frame_length=frame_length,
+            hop_length=hop_length)
+
+        def maybe_resample(y_blocks):
+            # will throw an error if audio is not valid
+            # resample if we have a different sr
+            for y in y_blocks:
+                librosa.util.valid_audio(y, mono=True)
+                if sr != orig_sr:
+                    y = librosa.resample(y, orig_sr, sr)
+                yield y
+
+        y_blocks = maybe_resample(y_blocks)
+
+    else:
+        if y is None or sr is None:
+            raise ParameterError(
+                'At least one of (y, sr) or filename must be provided')
+
+        librosa.util.valid_audio(y, mono=True)
+
+        # get block length, remove frame overhang (reuses conversions which does flooring)
+        block_length = max(segment_duration * sr, frame_length) * n_blocks # min block size = 1 frame
+        block_length = sample2frame(block_length, frame_length, hop_length) # drop overhang
+        block_length = frame2sample(block_length, frame_length, hop_length) # convert back
+
+        # get frames from array
+        y_blocks = librosa.util.frame(y, block_length, block_length).T
+
+    if full_frames: # drop any frames that are incomplete
```

Comment:
i don't think there is a reason to support this. dropping data is rarely what the user wants. and if they want it, let them take responsibility for it.
however, we may need better padding abstractions (at the frame level, not at the block level)

Reply:
I actually did a bad job at explaining this. I only just realized.

We're not actually dropping any samples here. We're just adjusting the block_length so that the block won't have incomplete frames.

`librosa.stream` takes `block_length` as a multiple of `frame_length` so in order to make `frame` work the same as `stream`, I made `block_length` a multiple of `frame_length`, which is really what I meant when I said "drop overhang".

Now that I explain it, I'm not sure I'm doing it properly. Need to look at librosa docs and double check.